### PR TITLE
Fix #40 to support rest client

### LINKED ISSUE #40: Rest client support
Are there any plans to add rest client support？

Quarkus provides two types of rest services [RESTEasy Classic](https://quarkus.io/guides/resteasy) and [REST](https://quarkus.io/guides/rest) which are implemented differently and have different dependencies.

And quarkus provide the differently rest client  implementation [resteasy-client](https://quarkus.io/guides/resteasy-client) and [rest-client](https://quarkus.io/guides/rest-client)

**Implementations and dependencies are also different  for different serializations**
[rest core](https://github.com/quarkusio/quarkus/tree/main/extensions/resteasy-reactive)
[resteasy classic core](https://github.com/quarkusio/quarkus/tree/main/extensions/resteasy-classic)
**For example**
_the jsonb serialization implementation for [REST](https://quarkus.io/guides/rest) and [rest-client](https://quarkus.io/guides/rest-client):_
```
implementation 'io.quarkus:quarkus-rest'
implementation 'io.quarkus:quarkus-rest-jsonb'
implementation 'io.quarkus:quarkus-rest-client-jsonb'
```

_the jsonb serialization implementation for [RESTEasy Classic](https://quarkus.io/guides/resteasy)  and [resteasy-client](https://quarkus.io/guides/resteasy-client):_
```
implementation 'io.quarkus:quarkus-resteasy'
implementation 'io.quarkus:quarkus-resteasy-jsonb'
implementation 'io.quarkus:quarkus-resteasy-client-jsonb'
```

### PR DESCRIPTION
**Pull request by zhfeng**
I have to disable the native test on `testClient` currently and will open an issue to investigate it.

### CODE REVIEW

**Comment by zhfeng** [comment]
The classic-resteasy-client is failing with
```
jakarta.ws.rs.ProcessingException: RESTEASY004655: Unable to invoke request: jakarta.ws.rs.ProcessingException: RESTEASY003215: could not find writer for content-type application/fury type: io.quarkiverse.fury.it.Bar
	at org.jboss.resteasy.client.jaxrs.engines.ManualClosingApacheHttpClient43Engine.invoke(ManualClosingApacheHttpClient43Engine.java:365)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:427)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:134)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:103)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:102)
	at jdk.proxy2/jdk.proxy2.$Proxy101.bar(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.jboss.resteasy.microprofile.client.ProxyInvocationHandler.invoke(ProxyInvocationHandler.java:152)
	at jdk.proxy2/jdk.proxy2.$Proxy102.bar(Unknown Source)
	at io.quarkiverse.fury.it.FuryTest.testRestClient(FuryTest.java:61)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:806)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: jakarta.ws.rs.ProcessingException: RESTEASY003215: could not find writer for content-type application/fury type: io.quarkiverse.fury.it.Bar
	at org.jboss.resteasy.core.interception.jaxrs.ClientWriterInterceptorContext.throwWriterNotFoundException(ClientWriterInterceptorContext.java:48)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.getWriter(AbstractWriterInterceptorContext.java:278)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.syncProceed(AbstractWriterInterceptorContext.java:219)
	at org.jboss.resteasy.core.interception.jaxrs.AbstractWriterInterceptorContext.proceed(AbstractWriterInterceptorContext.java:207)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.writeRequestBody(ClientInvocation.java:387)
	at org.jboss.resteasy.client.jaxrs.engines.ManualClosingApacheHttpClient43Engine.writeRequestBodyToOutputStream(ManualClosingApacheHttpClient43Engine.java:579)
	at org.jboss.resteasy.client.jaxrs.engines.ManualClosingApacheHttpClient43Engine.buildEntity(ManualClosingApacheHttpClient43Engine.java:548)
	at org.jboss.resteasy.client.jaxrs.engines.ManualClosingApacheHttpClient43Engine.loadHttpMethod(ManualClosingApacheHttpClient43Engine.java:486)
	at org.jboss.resteasy.client.jaxrs.engines.ManualClosingApacheHttpClient43Engine.invoke(ManualClosingApacheHttpClient43Engine.java:345)
	... 13 more

```

I tried to add `FurySerializer` in the reflective classes but it seems not work.

The reactive-rest-client is failing with
```
java.lang.IllegalStateException: No RestClientBuilderResolver implementation found!
	at org.eclipse.microprofile.rest.client.spi.RestClientBuilderResolver.instance(RestClientBuilderResolver.java:81)
	at org.eclipse.microprofile.rest.client.RestClientBuilder.newBuilder(RestClientBuilder.java:54)
	at io.quarkiverse.fury.it.FuryTest.testRestClient(FuryTest.java:59)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestMethod(QuarkusTestExtension.java:806)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

I'm not sure how to fix it.

@geoand any hint?

File: deployment/src/main/java/io/quarkiverse/fury/deployment/FuryProcessor.java [hunk 1]
```diff
@@ -86,20 +83,29 @@ public void registerResteasyIntegration(Capabilities capabilities,
...
+        additionalReaders.produce(new MessageBodyReaderBuildItem.Builder(
+                ReactiveFurySerializer.class.getName(), Object.class.getName())
+                .setMediaTypeStrings(List.of("application/fury", "application/*+fury"))
+                .setRuntimeType(type)
+                .setBuiltin(true).build());
+        additionalWriters.produce(new MessageBodyWriterBuildItem.Builder(
+                ReactiveFurySerializer.class.getName(), Object.class.getName())
+                .setMediaTypeStrings(List.of("application/fury", "application/*+fury"))
+                .setRuntimeType(type)
+                .setBuiltin(true).build());
```
Comment (geoand) [inline deployment/src/main/java/io/quarkiverse/fury/deployment/FuryProcessor.java:109]: `ReactiveFurySerializer` is a server part, so it won't work for the client

File: deployment/src/main/java/io/quarkiverse/fury/deployment/FuryProcessor.java [hunk 2]
```diff
@@ -86,20 +83,29 @@ public void registerResteasyIntegration(Capabilities capabilities,
             BuildProducer<ResteasyJaxrsProviderBuildItem> resteasyJaxrsProviderBuildItemBuildProducer,
             BuildProducer<MessageBodyReaderBuildItem> additionalReaders,
             BuildProducer<MessageBodyWriterBuildItem> additionalWriters) {
-        if (capabilities.isPresent(Capability.RESTEASY)) {
+        if (capabilities.isPresent(Capability.RESTEASY) || capabilities.isPresent(Capability.RESTEASY_CLIENT)) {
             resteasyJaxrsProviderBuildItemBuildProducer
                     .produce(new ResteasyJaxrsProviderBuildItem(FurySerializer.class.getName()));
```
Comment (geoand) [inline deployment/src/main/java/io/quarkiverse/fury/deployment/FuryProcessor.java:88]: I am not sure if this is enough for RESTEasy Classic. You probable also have to use `META-INF/services/jakarta.ws.rs.ext.Providers` like RESTEasy does in the `resteasy-json-binding-provider` module for example

File: deployment/src/main/java/io/quarkiverse/fury/deployment/FuryProcessor.java (hunk 1 above)
Comment (geoand) [reply deployment/src/main/java/io/quarkiverse/fury/deployment/FuryProcessor.java:109]: Oh, I didn't see that

**Pull request merged**